docs/design/daemonless.md

```diff
@@ -0,0 +1,139 @@
+> and handle this in bpfd itself.
+
+### Command: bpfd load file
+
+As the name suggests, this command will load an eBPF program from a file.
+This was formerly `bpfctl load-from-file`.
+
+### Command: bpfd load image
+
+As the name suggests, this command will load an eBPF program from a container
+image. This was formerly `bpfctl load-from-image`.
+
+### Command: bpfd unload
+
+This command will unload an eBPF program. This was formerly `bpfctl unload`.
+
+### Command: bpfd list
+
+This command will list the eBPF programs that are currently loaded.
+This was formerly `bpfctl list`.
+
+### Command: bpfd pull
+
+This command will pull the bpfd container image from a registry.
+This was formerly `bpfctl pull`.
+
+### Command: bpfd images
+
+This command will list the bpfd container images that are available.
+This command didn't exist, but makes sense to add.
+
+### Command: bpfd version
+
+This command will print the version of bpfd.
+This command didn't exist, but makes sense to add.
+
+### State Management
+
+This is perhaps the most significant change from how bpfd currently works.
+
+Currently bpfd maintains state in-memory about the programs that have been
+loaded (by bpfd, and the kernel).
+Some of this state is flushed to disk, so if bpfd is restarted it can
+reconstruct it.
+
+Flushing to disk and state reconstruction is cumbersome at present and having
+to move all state management out of in-memory stores is a forcing function
+to improve this.
+We will replace the existing state management with [sled], which gives us
+a familiar API to work with while also being fast, reliable and persistent.
+
+[sled]: https://github.com/spacejam/sled
+
+### Metrics and Monitoring
+
+While adding metrics and monitoring is not a goal of this design, it should
+nevertheless be a consideration. In order to provide metrics to Prometheus
+or OpenTelemetry we will require an additional exporter process.
+
+We can either:
+
+1. Use the bpfd socket and retrieve metrics via the gRPC API
+1. Place state access + metrics gathering functions in a library, such that
+   they could be used directly by the exporter process without requiring the
+   bpfd socket.
+
+The latter would be more inline with how podman-prometheus-exporter works.
+The benefit here is that, the metrics exporter process can be long running
+with less privileges - whereas if it were to hit the API over the socket it
+would effectively negate the point of being daemonless in the first place since
+collection will likley occur more frequently than the timeout on the socket.
+
+## Benefits
+
+The benefits of this design are:
+
+- No long-running daemon with root privileges
+- No need to run as a non-root user, this is important since the number of
+  capabilities required is only getting larger.
+- We only need to ship a single binary
+- We can use systemd socket activation to start the daemon on demand
+- Better state management
```

Comment:
We could also introduce this new state management method with the daemon....right?

Reply:
You could, but the onus is on us to keep state out of memory.
In the daemonless case we literally have no choice so it can't creep back in.